rustls/src/server/mod.rs

```diff
@@ -146,6 +146,311 @@ impl<'a> ClientHello<'a> {
+    pub fn with_single_cert_with_ocsp_and_sct(
+        self,
+        cert_chain: Vec<key::Certificate>,
+        key_der: key::PrivateKey,
+        ocsp: Vec<u8>,
+        scts: Vec<u8>,
+    ) -> Result<ServerConfigBuilderComplete, TlsError> {
+        let resolver =
+            handy::AlwaysResolvesChain::new_with_extras(cert_chain, &key_der, ocsp, scts)?;
+        Ok(self.with_cert_resolver(Arc::new(resolver)))
+    }
+
+    /// Sets a custom `ResolvesServerCert`.
+    pub fn with_cert_resolver(
+        self,
+        resolver: Arc<dyn ResolvesServerCert>,
+    ) -> ServerConfigBuilderComplete {
+        ServerConfigBuilderComplete {
+            ciphersuites: self.ciphersuites,
+            kx_groups: self.kx_groups,
+            verifier: self.verifier,
+            cert_resolver: resolver,
+            optional: Default::default(),
+        }
+    }
+}
+
+/// Items in ServerConfig where we hope to provide good, generally-applicable
+/// defaults.
+///
+/// Names and meanings match ServerConfig.
+struct ServerConfigOptions {
+    ignore_client_order: bool,
+    mtu: Option<usize>,
+    session_storage: Arc<dyn StoresServerSessions + Send + Sync>,
+    ticketer: Arc<dyn ProducesTickets>,
+    alpn_protocols: Vec<Vec<u8>>,
+    versions: Vec<ProtocolVersion>,
+    key_log: Arc<dyn KeyLog>,
+}
+
+impl Default for ServerConfigOptions {
+    fn default() -> Self {
+        Self {
+            ignore_client_order: false,
+            mtu: None,
+            session_storage: handy::ServerSessionMemoryCache::new(256),
+            ticketer: Arc::new(handy::NeverProducesTickets {}),
+            alpn_protocols: Vec::new(),
+            versions: vec![ProtocolVersion::TLSv1_3, ProtocolVersion::TLSv1_2],
+            key_log: Arc::new(NoKeyLog {}),
+        }
+    }
+}
+
+/// A ServerConfigBuilder where we know all mandatory configuration.
+pub struct ServerConfigBuilderComplete {
+    ciphersuites: Vec<&'static SupportedCipherSuite>,
+    kx_groups: Vec<&'static SupportedKxGroup>,
+    verifier: Arc<dyn verify::ClientCertVerifier>,
+    cert_resolver: Arc<dyn ResolvesServerCert>,
+    optional: ServerConfigOptions,
+}
+
+impl ServerConfigBuilderComplete {
+    pub fn build(self) -> ServerConfig {
+        ServerConfig {
+            ciphersuites: self.ciphersuites,
+            kx_groups: self.kx_groups,
+            ignore_client_order: self.optional.ignore_client_order,
+            mtu: self.optional.mtu,
+            session_storage: self.optional.session_storage,
+            ticketer: self.optional.ticketer,
+            cert_resolver: self.cert_resolver,
+            alpn_protocols: self.optional.alpn_protocols,
+            versions: self.optional.versions,
+            verifier: self.verifier,
+            key_log: self.optional.key_log,
+            #[cfg(feature = "quic")]
+            max_early_data_size: 0,
+        }
+    }
+
+    /// Set the MTU to `mtu`.
+    pub fn set_mtu(&mut self, mtu: Option<usize>) {
```

Comment:
I think these whould probably lose the `set_` prefix and either take `self` and return `Self` or return `&mut Self`, so that the more fluent combinator style can be used at this stage, too.

Reply:
The naming convention without `set_` does seem to be what hyper and other libraries that I'm familiar with do.